ProjFS.Mac/PrjFSLib/PrjFSUser.cpp

```diff
@@ -11,13 +11,51 @@ struct DarwinVersion
     unsigned long major, minor, revision;
 };
 
+struct PrjFSService_WatchContext
+{
+    std::function<void(io_service_t, io_connect_t, PrjFSService_WatchContext*)> discoveryCallback;
+    io_iterator_t notificationIterator;
+    PrjFSServiceUserClientType clientType;
+};
+
+
 typedef decltype(IODataQueueDequeue)* ioDataQueueDequeueFunctionPtr;
 static ioDataQueueDequeueFunctionPtr ioDataQueueDequeueFunction = nullptr;
 typedef decltype(IODataQueuePeek)* ioDataQueuePeekFunctionPtr;
 static ioDataQueuePeekFunctionPtr ioDataQueuePeekFunction = nullptr;
 
 static void InitDataQueueFunctions();
 
+bool PrjFSService_ValidateVersion(io_service_t prjfsService)
+{
+    CFTypeRef kextVersionObj = IORegistryEntryCreateCFProperty(prjfsService, CFSTR(PrjFSKextVersionKey), kCFAllocatorDefault, 0);
+    CFStringRef kextVersionString;
+    if (nullptr == kextVersionObj || CFStringGetTypeID() != CFGetTypeID(kextVersionObj))
+    {
+        std::cerr << "PrjFS kernel service does not advertise valid kext version property\n";
+        goto CleanupAndFail;
```

Comment:
curious why we favor goto statements rather than a function call here

Reply:
Here, there isn't much to clean up, but in more complex functions, the number of variables you'd have to pass to a cleanup helper function could rapidly become annoying. Plus, pushing that logic out to another function is probably visually less appealing, and could encourage failing to update the cleanup when changing the core function logic.

If anything, the "modern C++" approach here would be to have [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization)-wrapper classes for this sort of thing.

Ultimately I guess going with gotos for cleanup/errors is mostly a stylistic choice that was decided early on during the Mac prototyping phase. The idea was, especially in the kext, to write C-like code with some hand-picked C++ features. (For example, C++ exception handling isn't supported by the compiler/runtime in the kext. C++ exceptions are also regarded by many C++ programmers as a bad idea that should never have been implemented, but I digress.) I think we agreed to try and nevertheless keep the user space code, where we're less constrained, more or less aligned with the kext coding style. Or at least that's roughly how I remember it. :-)